smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/AuthSchemeResolverGenerator.kt

```diff
@@ -0,0 +1,254 @@
+                        opShape.hasTrait(UnsignedPayloadTrait::class.java)
+                    ) {
+                        val opName = it.name.toLowerCamelCase()
+                        val sdkId = getSdkId(ctx)
+                        val validSchemesForOp = serviceIndex.getEffectiveAuthSchemes(
+                            ctx.service,
+                            it,
+                            ServiceIndex.AuthSchemeMode.NO_AUTH_AWARE
+                        )
+                        renderOperationSwitchCase(sdkId, opShape, opName, validSchemesForOp, writer)
+                    }
+                }
+                // Handle default case, where operations default to auth schemes defined on service shape
+                val validSchemesForService = serviceIndex.getEffectiveAuthSchemes(ctx.service, ServiceIndex.AuthSchemeMode.NO_AUTH_AWARE)
+                renderDefaultSwitchCase(getSdkId(ctx), validSchemesForService, writer)
+            }
+
+            // Return result
+            write("return validAuthOptions")
+        }
+    }
+
+    private fun renderOperationSwitchCase(sdkId: String, opShape: OperationShape, opName: String, schemes: Map<ShapeId, Trait>, writer: SwiftWriter) {
+        writer.apply {
+            write("case \"$opName\":")
+            indent()
+            schemes.forEach {
+                if (it.key == SigV4Trait.ID) {
+                    write("var sigV4Option = AuthOption(schemeID: \"${it.key}\")")
+                    write("sigV4Option.signingProperties.set(key: AttributeKeys.signingName, value: ${(it.value as SigV4Trait).name})")
+                    openBlock("guard let region = serviceParams.region else {", "}") {
+                        val errorMessage = "\"Missing region in auth scheme parameters for SigV4 auth scheme.\""
+                        write("throw ClientError.authError($errorMessage)")
+                    }
+                    write("sigV4Option.signingProperties.set(key: AttributeKeys.signingRegion, value: region)")
+
+                    val unsignedBody = opShape.hasTrait(UnsignedPayloadTrait::class.java)
+                    val signedBodyHeader = if ((sdkId == "s3" || sdkId == "glacier") && !unsignedBody) ".contentSha256" else ".none"
+                    // Set .unsignedBody to true IFF operation has UnsignedPayloadTrait
+                    write("sigV4Option.signingProperties.set(key: AttributeKeys.unsignedBody, value: $unsignedBody)")
+                    // Set .signedBodyHeader to .contentSha256 IFF service is S3 / Glacier AND operation does not have UnsignedPayloadTrait.
+                    // Set to .none otherwise.
+                    write("sigV4Option.signingProperties.set(key: AttributeKeys.signedBodyHeader, value: $signedBodyHeader)")
+
+                    write("validAuthOptions.append(sigV4Option)")
+                } else {
+                    write("validAuthOptions.append(AuthOption(schemeID: \"${it.key}\"))")
+                }
```

Comment:
It looks like there is a lot of code duplicated between `renderOperationSwitchCase` and `renderDefaultSwitchCase`, what do you think about writing a helper that is re-used in both?

Reply:
Ok so a lot of the logic here is re-located to AWSClientRuntime (to reduce redundant codegen & to have ONE place that has all the signingProperties logic) in the next milestone (customizations milestone) and there this is refactored. I could refactor the method shown here, but that would be pointless since the logic should be changed in next set of PRs & refactor is done there.

Reply:
No need to refactor here if you have plans to clean this up in the next milestone, let's just make sure we do that